src/com/lld/inmemorycache/model/impl/InMemoryStorage.java

```diff
@@ -0,0 +1,66 @@
+package com.lld.inmemorycache.model.impl;
+
+import com.lld.inmemorycache.model.IStorage;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class InMemoryStorage implements IStorage {
+    private ConcurrentHashMap<String, String> _storage;
+    private static ReentrantLock lock;
+
+    public InMemoryStorage() {
+        _storage = new ConcurrentHashMap<>();
+        // fairness: first come, first served.
+        lock = new ReentrantLock(true);
+    }
+
+    @Override
+    public boolean put(String key, String value) {
+        lock.lock();

```

Comment:
Since storage is used by the cache along with eviction policy, taking individual locks in the storage or eviction side alone won't work.